fix(ua): change the User-Agent to use the new specs lib (version)

robertmogos opened the pull request:
<!--
  Thanks for submitting a pull request!
  Please provide enough information so that others can review your pull request.
-->

**Summary**
Update UA to `angular ($VERSION); angular-instantsearch ($VERSION)`

Add the `addAlgoliaAgent` to `SearchClient` and fix the issue where the UA is **only** added when the `SearchClient` is not passed in config

<!--
  Explain the **motivation** for making this change.
  What existing problem does the pull request solve?
  Are there any linked issues?
-->

**Result**

<!--
  Demonstrate the code is solid.
  Example: The exact commands you ran and their output,
  screenshots / videos if the pull request changes UI.
-->

francoischalifour commented on src/create-ssr-algolia-client.ts:
```diff
@@ -39,7 +39,7 @@ export function createSSRSearchClient({
 }) {
   const client = algoliasearch(appId, apiKey, {});
-  client.addAlgoliaAgent(`angular-instantsearch ${VERSION}`);
+  client.addAlgoliaAgent(`angular-instantsearch (${VERSION})`);
```
We can add a mention that it's on the server.

See https://github.com/algolia/react-instantsearch/pull/2209#pullrequestreview-217245480.

samouss commented on src/create-ssr-algolia-client.ts:
```diff
@@ -39,7 +39,7 @@ export function createSSRSearchClient({
 }) {
   const client = algoliasearch(appId, apiKey, {});
-  client.addAlgoliaAgent(`angular-instantsearch ${VERSION}`);
+  client.addAlgoliaAgent(`angular-instantsearch (${VERSION})`);
```
We can do the same than for the React version have a dedicated user agent for the server.

robertmogos replied:
@samouss @francoischalifour  `angular-instantsearch-server` looks good?

samouss replied:
Yes, it's the same version.

francoischalifour replied:
Do we want to track the Node version?

samouss replied:
Not sure it's really useful in that case since we are bound to the Angular support of Node.

Haroenv replied:
the node version itself could be indicated by the JS client as noted by @samouss in the client PR

robertmogos replied:
Done

francoischalifour commented:
Before approving, this should be tested for both client and server versions.

robertmogos commented:
> Before approving, this should be tested for both client and server versions.

@francoischalifour Tested both on client and server:

Netlify:
`angular (7.0.1);angular-instantsearch (2.2.1)`

SSR:
`angular-instantsearch (2.2.1);angular-instantsearch-server (2.2.1);`

Haroenv commented:
looks like SSR is missing the Angular version then?

robertmogos commented:
> looks like SSR is missing the Angular version then?

@Haroenv I didn't added it since in create-ssr-algolia-client.ts there is no import related to Angular. Should I add it ? 

```
import * as algoliasearchProxy from 'algoliasearch/index';
import * as encodeProxy from 'querystring-es3/encode';
import { VERSION } from './version';
```

Haroenv commented:
It would seem useful to know for me: 

```ts
import {
   VERSION as AngularVersion,
 } from '@angular/core';
```

Haroenv commented:
Looks like prettier wants the imports inlined. Also there are no tests for this, and they probably should be added

robertmogos commented:
> It doesn't seem to have a test for the client side.

I added the client side UA + thanks to @tkrugg we spot a bug where the UA was only added when the `SearchClient` was **not** provided in the config. I added too the `addAlgoliaAgent: (agent: string) => void;` method to the `SearchClient`

francoischalifour commented on src/instantsearch/instantsearch.ts:
```diff
@@ -262,12 +264,12 @@ export class NgAisInstantSearch implements AfterViewInit, OnInit, OnDestroy {
       config.appId = undefined;
       config.apiKey = undefined;
     }
+    config.searchClient.addAlgoliaAgent(`angular (${AngularVersion.full})`);
```
This will break with a custom `searchClient` that doesn't expose the `addAlgoliaAgent` function (we can add a specific test for that).

francoischalifour commented on src/instantsearch/instantsearch.ts:
```diff
@@ -262,12 +264,12 @@ export class NgAisInstantSearch implements AfterViewInit, OnInit, OnDestroy {
     // custom algolia client agent
     if (!config.searchClient && !config.createAlgoliaClient) {
```
As explained in private, this check doesn't make sense anymore because we now recommend using the `searchClient` with our official JS client (instead of using `appId` and `apiKey`).

We should change this check to something like:

```js
config.searchClient = config.searchClient || algoliasearch(config.appId, config.apiKey);

if (typeof config.searchClient.addAlgoliaAgent === 'function') {
  // add user agents
}
```

`createAlgoliaClient` is deprecated and should be removed in a next major btw (it's replaced by `searchClient`). So are `appId` and `apiKey`. We can add a comment to mention that.

francoischalifour commented on src/instantsearch/instantsearch.ts:
```diff
@@ -146,6 +147,7 @@ export type SearchForFacetValuesResponse = {
 };
 
 export type SearchClient = {
+  addAlgoliaAgent: (agent: string) => void;
```
It doesn't necessarily exist if it's a custom search client.

```suggestion
  addAlgoliaAgent?: (agent: string) => void;
```

samouss commented on src/instantsearch/__tests__/instantsearch.spec.ts:
```diff
@@ -0,0 +1,46 @@
+    expect(addAlgoliaAgent).toHaveBeenCalledWith(
+      `angular-instantsearch (${VERSION})`
+    );
+  });
```
We can add a test on with a `searchClient` that does not implement `addAlgoliaAgent`.

samouss commented on src/instantsearch/__tests__/instantsearch.spec.ts:
```diff
@@ -0,0 +1,46 @@
+jest.mock('../../../src/base-widget');
+
+describe('InstantSearch', () => {
+  it('should add user agent stuff when the client is provided to the config', () => {
```
```suggestion
  it('should add user agents when the client is provided to the config', () => {
```

francoischalifour commented on src/instantsearch/instantsearch.ts:
```diff
@@ -260,15 +262,14 @@ export class NgAisInstantSearch implements AfterViewInit, OnInit, OnDestroy {
-    if (!config.searchClient && !config.createAlgoliaClient) {
-      const client = algoliasearch(config.appId, config.apiKey);
-      client.addAlgoliaAgent(`angular-instantsearch ${VERSION}`);
+    config.searchClient =
```
The [storybook breaks](https://deploy-preview-481--angular-instantsearch.netlify.com/examples/storybook/?selectedKind=ClearRefinements&selectedStory=default&full=0&addons=1&stories=1&panelRight=0) because we need to remove `appId` and `apiKey` from the configuration before passing it to the InstantSearch.js instantiation.

If we want to release this feature before releasing v3, we actually might want to revert this change and keep the `if (typeof config.searchClient.addAlgoliaAgent === 'function')` statement. It might break some people's implementation who still rely on `createAlgoliaClient`.

francoischalifour commented on src/instantsearch/__tests__/instantsearch.spec.ts:
```diff
@@ -0,0 +1,77 @@
+    );
+  });
+
+  it('should not add a user agent when the client is not provided to the config', () => {
```
Do you rather mean this?

```suggestion
  it('should not add a user agent when addAlgoliaAgent is not provided in the client', () => {
```
